src/orpheus_model.cpp

```diff
@@ -0,0 +1,461 @@
+    
+    for (int l = 0; l < model->n_layers; l++) {
+        struct ggml_tensor * residual = inpL;
+        cur = orpheus_build_layer_norm(ctx, inpL, model->layers[l].input_norm);
+
+        struct ggml_tensor * attn_out;
+
+        // self-attention
+        {
+            struct ggml_tensor * Qcur = ggml_mul_mat(ctx, model->layers[l].q, cur);
+            struct ggml_tensor * Kcur = ggml_mul_mat(ctx, model->layers[l].k, cur);
+            struct ggml_tensor * Vcur = ggml_mul_mat(ctx, model->layers[l].v, cur);
+
+            orpheus_build_kv_store(ctx, gf, Kcur, Vcur, l, batch.n_tokens, 3);
+            struct ggml_tensor * k =
+                ggml_cont(ctx, ggml_view_3d(ctx, kv_self->k_l[l],
+                        model->head_size, full_sequence_length, model->n_attn_heads,
+                        ggml_element_size(kv_self->k_l[l]) * model->n_attn_heads * model->head_size,
+                        ggml_element_size(kv_self->k_l[l]) * model->head_size,
+                        0));            
+            
+            struct ggml_tensor * v =
+                ggml_view_2d(ctx, kv_self->v_l[l],
+                        model->hidden_size, full_sequence_length,
+                        ggml_element_size(kv_self->k_l[l]) * model->hidden_size,
+                        0);
+
+            v = ggml_cont_3d(ctx, ggml_transpose(ctx, v), full_sequence_length, model->head_size, model->n_attn_heads);
+
+            Qcur = ggml_rope_ext(
+                ctx, ggml_cont(ctx, ggml_reshape_3d(ctx, Qcur, model->head_size, model->n_attn_heads, batch.n_tokens)), 
+                octx->positions, model->rope_frequencies, model->head_size, 2, 0, 500000.0f, // rope theta
+                1.0f, 0.0f, 1.0f, 0.0f, 0.0f);
+
+            struct ggml_tensor * q = ggml_cont(ctx, ggml_permute(ctx, Qcur, 0, 2, 1, 3));
+            struct ggml_tensor * kq = ggml_mul_mat(ctx, k, q);
+            kq = ggml_soft_max_ext(ctx, kq, KQ_mask_dec, 1.0f/sqrtf(model->head_size), 0.0f);
+            struct ggml_tensor * kqv = ggml_mul_mat(ctx, kq, v);
+            struct ggml_tensor * kqv_merged = ggml_permute(ctx, kqv, 2, 0, 1, 3);
+            attn_out = ggml_cont_2d(ctx, kqv_merged, model->hidden_size, batch.n_tokens);
+            attn_out = ggml_mul_mat(ctx, model->layers[l].o, attn_out);
+        }
+
+        cur = ggml_add(ctx, attn_out, residual);
+        
+        struct ggml_tensor * residualffn = cur;
+
+        // mlp
+        {
+            cur = orpheus_build_layer_norm(ctx, cur, model->layers[l].post_attention_norm);
+            cur = ggml_mul(ctx, ggml_silu(ctx, ggml_mul_mat(ctx, model->layers[l].gate, cur)), ggml_mul_mat(ctx, model->layers[l].up, cur));
+            cur = ggml_mul_mat(ctx, model->layers[l].down, cur);
+        }
+        cur = ggml_add(ctx, cur, residualffn);
+        inpL = cur;
+    }
+    
+    cur = orpheus_build_layer_norm(ctx, cur, model->output_norm);
+    cur = ggml_mul_mat(ctx, model->head, cur);
```

Comment:
We should comment somewhere that this tensor (as Orpheus has it implemented) is inefficient and unnecessarily big. Generation only uses ~40k of the ~160k output space.